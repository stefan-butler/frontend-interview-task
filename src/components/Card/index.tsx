import { Allowance } from '@/types/allowances'
interface CardProps {
  allowance: Allowance
}

const Card = ({ allowance }: CardProps) => {
  return (
    <div className="border border-gray-200 shadow-sm rounded-lg w-[380px] h-[190px] flex flex-col justify-between">
      <div className="flex-grow p-5">
        <h3 className="text-lg ">{allowance.name}</h3>
        {allowance.type === 'card' ? (
          <p className="text-sm text-gray-300 font-light">Spend Card</p>
        ) : (
          <p className="text-sm text-gray-300 font-light">Expense</p>
        )}
      </div>
      {allowance.active && (
        <div className="flex-grow px-5 mt-9">
          <div className="flex justify-between text-sm mb-1">
            <p>
              {Math.round(
                (Number(allowance.spent) / Number(allowance.amount)) * 100
              )}
              % utilised
            </p>
            <p className="text-gray-300">
              {allowance.currency}
              {Number(allowance.amount).toFixed(2)} /{' '}
              {allowance.renewal.charAt(0).toUpperCase() +
                allowance.renewal.slice(1)}
            </p>
          </div>
          <div className="w-full bg-gray-200 h-1 rounded-lg">
            <div
              className="bg-lime-400 h-1 rounded-lg"
              style={{
                width: `${Math.round((Number(allowance.spent) / Number(allowance.amount)) * 100)}%`,
              }}
            />
          </div>
        </div>
      )}
      {!allowance.active && (
        <div className="bg-gray-100 flex flex-grow justify-start items-center w-full h-5 rounded-lg">
          <p className="text-lime-500 px-5">Activate card</p>
        </div>
      )}
    </div>
  )
}

export default Card
