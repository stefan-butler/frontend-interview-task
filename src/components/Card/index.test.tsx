import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import Card from '.'
import data from '@/data/data.json'
import { Allowance } from '@/types/allowances'

describe('Card', () => {
  let mockAllowance: Allowance

  beforeEach(() => {
    mockAllowance = data[0]
  })

  it('renders as expected', () => {
    const { container } = render(<Card allowance={mockAllowance} />)
    expect(container).toMatchSnapshot()
  })

  it('displays the allowance name correctly', () => {
    render(<Card allowance={mockAllowance} />)
    expect(screen.getByText(mockAllowance.name)).toBeInTheDocument()
  })

  it('calculates and displays percentage correctly', () => {
    render(<Card allowance={mockAllowance} />)
    const activeCard = { ...mockAllowance, active: true }
    const percentage = Math.round(
      (Number(activeCard.spent) / Number(activeCard.amount)) * 100
    )
    expect(screen.getByText(`${percentage}% utilised`)).toBeInTheDocument()
  })

  it('renders "Activate Card" when allowance is inactive', () => {
    const inactiveAllowance = { ...mockAllowance, active: false }
    render(<Card allowance={inactiveAllowance} />)
    expect(screen.getByText('Activate card')).toBeInTheDocument()
  })
})
