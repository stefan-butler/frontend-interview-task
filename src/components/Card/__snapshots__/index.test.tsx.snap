// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders as expected 1`] = `
<div>
  <div
    class="border border-gray-200 shadow-sm rounded-lg w-[380px] h-[190px] flex flex-col justify-between"
  >
    <div
      class="flex-grow p-5"
    >
      <h3
        class="text-lg "
      >
        Learning & Development
      </h3>
      <p
        class="text-sm text-gray-300 font-light"
      >
        Expense
      </p>
    </div>
    <div
      class="flex-grow px-5 mt-9"
    >
      <div
        class="flex justify-between text-sm mb-1"
      >
        <p>
          20
          % utilised
        </p>
        <p
          class="text-gray-300"
        >
          Â£
          100.00
           /
           
          Month
        </p>
      </div>
      <div
        class="w-full bg-gray-200 h-1 rounded-lg"
      >
        <div
          class="bg-lime-400 h-1 rounded-lg"
          style="width: 20%;"
        />
      </div>
    </div>
  </div>
</div>
`;
